### PR TITLE
ROX-24948: external entities API extension

### PR DESCRIPTION
**Pull request by Stringy**
### Description

This PR modifies and adds to the Network Graph in the following ways:

- Converts GetExternalEntities to use a query rather than a predicate. This simplifies the API endpoint and allows us to better take advantage of Postgresql. Combined with #12676 we can filter endpoints based on CIDR/subnet. e.g. get all external entities within 192.168.0.0/16
- Adds a new endpoint GetExternalNetworkFlows to retrieve all external network flows for a given deployment. The intended use case for now is API-only supplement to the external IP feature. A small change to the NetworkFlow store has been made to achieve this, again leveraging some of Postgres to make the queries quicker.

Depends on #11746 and #12676 

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [x] [documentation PR] We have a dedicated ticket for that: ROX-26639  

### Testing and quality

- [ ] ~~the change is production ready: the change is GA or otherwise~~ the functionality is gated by a feature flag
- [x] CI results are inspected

#### Automated testing

- [x] added unit tests
- [x] e2e tests (mostly covered for regression by the networkgraph tests)
- [ ] modified existing tests

#### How I validated my change

Primarily validated through manual testing through the use of a supplementary script: https://github.com/stackrox/contributions/pull/107 

We plan to enable the External-IPs feature-flag for all CI tests and update the networkgraph groovy test to turn-on and off at Collector level.

### CODE REVIEW

File: central/networkgraph/entity/datastore/datastore_impl.go
Comment (JoukoVirtanen) [inline central/networkgraph/entity/datastore/datastore_impl.go:236]: There should probably be some error handling here, even though there is error handling where it is called.

File: central/networkgraph/service/service_impl.go
Comment (JoukoVirtanen) [inline central/networkgraph/service/service_impl.go:104]: I don't like this abbreviation. Perhaps change `ets` to `entities`. In that case you should change `s.entities` to `s.entitiesDS`.

File: central/networkgraph/service/service_impl.go
Comment (JoukoVirtanen) [inline central/networkgraph/service/service_impl.go:138]: Save `flow.GetProps().GetSrcEntity()` and `flow.GetProps().GetDstEntity()` to variables. Also you probably need to check for nil.

File: central/networkgraph/entity/datastore/datastore_impl.go
Comment (dashrews78) [reply central/networkgraph/entity/datastore/datastore_impl.go:236]: Why?  Basically just a pass through function.

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (dashrews78) [reply central/networkgraph/flow/datastore/internal/store/postgres/store.go:58]: @JoukoVirtanen 
By and large we work on the serialized object because that is where ALL the data is.  Thus it is simpler to work on the serialized object than to get 4 pieces of data from columns and 10 pieces of data from the serialized object.  That was to maintain the previous pattern an minimize code change.  

Now all that said, network flows is a totally custom store and a partitioned table due to the amount of data and its access patterns.  So Network Flows doesn't even have a serialized object.

File: central/networkgraph/service/service_impl.go
Comment (Molter73) [reply central/networkgraph/service/service_impl.go:104]: You could also change it to ents, they might help you fight Sauron...

image

File: central/networkgraph/service/service_impl.go
Comment (Molter73) [inline central/networkgraph/service/service_impl.go:116]: Don't have enough knowledge about SAC, but just checking, could this be used to get the external entities of a deployment that I don't have access to directly? Or is that checked before we get here?

File: central/networkgraph/service/service_impl.go
Comment (Molter73) [reply central/networkgraph/service/service_impl.go:138]: +1 to the nil check, looks like `GetProps()`, `GetSrcEntity()` and `GetDstEntity()` can all return nil, so we probably want to check it unless we have a strong reason why we shouldn't.

File: proto/api/v1/network_graph_service.proto
Comment (Molter73) [inline proto/api/v1/network_graph_service.proto:76]: Why is deployment_id not 2? This feels very odd, I would understand it on an existing message, but this one looks brand new.

Comment (Stringy) [reply proto/api/v1/network_graph_service.proto:76]: I was sure I'd fixed that - apologies. There used to be a `query` parameter, but it was unused due to the changes I made to the flow store, so I dropped it. I'll fix this.

File: central/networkgraph/service/service_impl.go
Comment (Stringy) [reply central/networkgraph/service/service_impl.go:116]: I had put this in primarily to follow similar behaviour to the network graph construction, but I think you're right, it would provide access to everything in the cluster. I'll use `ctx` instead, which, if my understanding is correct, will have the access scope of the request (some local testing of this is working as expected)

Comment (Molter73) [reply central/networkgraph/service/service_impl.go]: Don't know about you, but I would greatly appreciate some input here from one of the SAC experts, I think that would be Merlin team? Should we ask them?

**Review by Molter73 — APPROVED** [review]
Changes LGTM, but I would still encourage getting reviews from people that have more knowledge about the code being modified before merging.

**Review by vikin91 — COMMENTED** [review]
I browsed the code and didn't see anything that would apply to the sensor part of this change. I could review the Central part, but I think there are colleagues who can do it better. 
But I like the change as it should make things easier when we have a lot of IPs in the DB :)

File: central/networkgraph/service/service.go
Comment (dashrews78) [inline central/networkgraph/service/service.go:50]: nit:  While making this change, it might be cleaner to update the parameter name too.

File: central/networkgraph/entity/datastore/datastore_impl_test.go
Comment (dashrews78) [inline central/networkgraph/entity/datastore/datastore_impl_test.go:229]: nit:  I recommend a test where you find nothing.  Just to be safe that the query builder and queries are doing what you expect them to do.

**Review by dashrews78 — APPROVED** [review]
Left a couple minor comments.

LGTM in general.

File: central/networkgraph/service/service.go
Comment (Stringy) [reply central/networkgraph/service/service.go]: Done

File: central/networkgraph/entity/datastore/datastore_impl_test.go
Comment (Stringy) [reply central/networkgraph/entity/datastore/datastore_impl_test.go:230]: Done

File: central/networkgraph/flow/datastore/flow_impl.go
Comment (rhybrillou) [inline central/networkgraph/flow/datastore/flow_impl.go:70]: I think this function misses access control checks (unfortunately, the network graph flow store is not a generated one, and the logic for result filtering is trickier than the one for common stores).

My take is that the access control would require two steps here:
1) get the target deployment (the flow datastore would need to access the deployment datastore)
```
  elevatedCtx := sac.WithGlobalAccessScopeChecker(
    ctx,
    sac.AllowFixedScopes(
      sac.AcccessModeScopeKeys(storage.Access_READ_ACCESS),
      sac.ResourceScopeKeys(resources.Deployment),
    ),
  )
  deployment, found, err := fds.deploymentDS.GetDeployment(elevatedCtx, deploymentID)
  if err != nil || !found {
    // If the requested source/target deployment does not exist, no need to pull any flow.
    return nil, err
  }
```
2) check the requester has `DeploymentExtension` (network graph) permission to the cluster/namespace where the target deployment is:
```
  allowed, err := networkGraphSAC.ReadAllowed(ctx, sac.KeyForNSScopedObj(deployment))
  if err != nil || !allowed {
    // if the requester is not allowed to see network graph data
    // in the namespace where the request deployment is, return
    // no data.
    return nil, err
  }
```

Both steps should be done before passing the request to the storage layer.

Comment (Stringy) [reply central/networkgraph/flow/datastore/flow_impl.go:70]: I tried to implement these checks at the service level ([here](https://github.com/stackrox/stackrox/pull/12689/files#diff-911ba7c3eb2a2c8dc360db82d65eec89fe9434edcfc7628e6da8c0d305811402R119-R135)) - is that enough or do I need to duplicate/move the checks to this level?

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (rhybrillou) [inline central/networkgraph/flow/datastore/internal/store/postgres/store.go:519]: Should a new operation metric be needed for that function (Get**External**FlowsForDeployment)

**Review by rhybrillou — APPROVED** [review]
Good to go from the Scoped Access Control perspective.

I'd split the changes to the `GetExternalNetworkEntities` service (and associated data storage pipeline) in a dedicated PR with more comprehensive testing on the replacement of predicate walk with GetByQuery.

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (dashrews78) [reply central/networkgraph/flow/datastore/internal/store/postgres/store.go:519]: IMO, a good point.  Though not super important so I would write a tech debt ticket and do it as follow up work eventually.

File: central/networkgraph/service/service_impl.go
Comment (ovalenti) [reply central/networkgraph/service/service_impl.go:111]: The cluster_id is part of the query URL, so I would expect that it is always set.

However, this made me realize that we are no more filtering on the cluster_id, and add the following lines to address this.

https://github.com/stackrox/stackrox/pull/12689/commits/811f41700d800e371cf87ff150cb97cf31a3a5c9

Is that ok @rhybrillou ?

File: central/networkgraph/service/service_impl.go
Comment (dashrews78) [inline central/networkgraph/service/service_impl.go:102]:
```suggestion
	// Retrieves entities where the cluster ID matches the request cluster OR where the cluster ID is empty indicating global entities.
```

File: central/networkgraph/service/service_impl.go
Comment (ovalenti) [reply central/networkgraph/service/service_impl.go:158]: Is it ok to go with this implementation for the tech-preview ?

File: central/networkgraph/service/service_impl.go
Comment (ovalenti) [reply central/networkgraph/service/service_impl.go]: This has been reviewed by @rhybrillou since then, and it seems that the current implementation of SAC is acceptable.

File: central/networkgraph/service/service_impl_test.go
Comment (ovalenti) [reply central/networkgraph/service/service_impl_test.go]: Updated as suggested

File: central/networkgraph/flow/datastore/internal/store/postgres/store.go
Comment (ovalenti) [reply central/networkgraph/flow/datastore/internal/store/postgres/store.go]: I have created a new metric as suggested at https://github.com/stackrox/stackrox/pull/12689/commits/ead68617ccaee0fa541dfcd467d52cd13786ba31

File: central/networkgraph/service/service_impl_test.go
Comment (ovalenti) [reply central/networkgraph/service/service_impl_test.go:202]: Ok, I add this to my cheat-sheet !

File: central/networkgraph/service/service_impl.go
Comment (rhybrillou) [reply central/networkgraph/service/service_impl.go:111]: with the query extension, this looks good.
thanks for the iteration here

File: central/networkgraph/service/service_impl.go
Comment (rhybrillou) [reply central/networkgraph/service/service_impl.go]: yes, please keep the suggested evolution in the backlog for post-mvp

Comment (ovalenti) [reply central/networkgraph/service/service_impl.go]: I created ROX-26627 to track this.

Comment (rhybrillou) [reply central/networkgraph/service/service_impl.go]: ROX-26627

**Comment by porridge** [comment]
Looks like this changed managed to break the style check somehow. First failure is on [its merge commit](https://github.com/stackrox/stackrox/commit/cde66f4cc404a00463842b2a2eaa73e06ee35212).

image

```
diff --git a/generated/api/v1/network_graph_service.swagger.json b/generated/api/v1/network_graph_service.swagger.json
index 37fe59314d..cadd67a3cd 100644
--- a/generated/api/v1/network_graph_service.swagger.json
+++ b/generated/api/v1/network_graph_service.swagger.json
@@ -496,7 +496,8 @@
         },
         "dstPort": {
           "type": "integer",
-          "format": "int64"
+          "format": "int64",
+          "description": "may be 0 if not applicable (e.g., icmp)."
         },
         "l4protocol": {
           "$ref": "#/definitions/storageL4Protocol"
```

cc @ovalenti

**Pull request merged**
